src/cipher.rs

```diff
@@ -557,6 +661,101 @@ impl rustls_client_cert_verifier {
     }
 }
 
+/// A builder for a `rustls_client_cert_verifier_optional`. This builder object can be used to configure
+/// certificate revocation lists, and then turned into a `rustls_client_cert_verifier_optional` once ready.
+pub struct rustls_client_cert_verifier_optional_builder {
+    _private: [u8; 0],
+}
+
+// XXX: contained value is consumed even on error, so this can contain None. but the caller
+// still needs to free it
+pub(crate) struct AllowAnyAnonymousOrAuthenticatedClientHolder(
+    Option<AllowAnyAnonymousOrAuthenticatedClient>,
+);
+
+impl CastPtr for rustls_client_cert_verifier_optional_builder {
+    type RustType = AllowAnyAnonymousOrAuthenticatedClientHolder;
+}
+
+impl BoxCastPtr for rustls_client_cert_verifier_optional_builder {}
+
+impl rustls_client_cert_verifier_optional_builder {
+    /// Create a new optional client certificate verifier builder for the root store.
```

Comment:
I'm finding that the adjectives here are hard to parse. Which is optional: the client certificate, the verifier, or the builder? I happen to know that the client certificate (which will be supplied by the handshake) is optional.

Looking back, probably the fault is in the naming of the existing types (client_cert_verifier_optional and client_cert_verifier). Since we're making breaking changes maybe we should align these with the names of the Rust types? E.g. `rustls_allow_any_authenticated_client_builder_new` and `rustls_allow_any_anonymous_or_authenticated_client_new`? A mouthful to be sure but has the advantage of consistency; and the matching comments might be a little easier to understand?

Also a nit: "using the root store" rather than "for the root store".

Reply:
>  Since we're making breaking changes maybe we should align these with the names of the Rust types?

That sounds good to me. I agree that the wordyness is a fair cost to pay for having easier to understand types that can be better mapped back to their Rustls equivalents. 

To try and keep this easy to review I broke the renaming into a separate commit ahead of introducing the builder type + CRL support.

> Also a nit: "using the root store" rather than "for the root store".

Fixed :+1: